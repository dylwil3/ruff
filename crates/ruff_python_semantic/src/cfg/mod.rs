//! # Overview
//! To build a control-flow graph, we step through each statement in order.
//! By default, statements are added to the current basic block until we
//! reach a statement that invokes control flow. These are exactly the following:
//!
//! | Branching                  | Loops   | Jumps      |
//! |----------------------------|---------|------------|
//! | `if`                       | `for`   | `break`    |
//! | `match`                    | `while` | `continue` |
//! | `try`                      |         | `raise`    |
//! | `with`                     |         | `return`   |
//!
//! There are also `assert` statements, which are equivalent
//! to a branch followed by a jump. That is,
//! ```
//! assert cond
//! ```
//!
//! is equivalent to
//!
//! ```
//! if not cond:
//!   raise AssertionError
//! ```
//!
//! (Technically there is an additional `if __debug__` wrapped around, but
//! we will ignore that.)
//!
//! The control flow graph is then constructed using this equivalent
//! form.
//!
//!
//! We now discuss how each kind of control flow is handled.
//!
//! ## Branching
//!
//! Upon reaching a branching statement, the statement
//! itself terminates the basic block, and several outgoing
//! edges are added, labeled by the condition that needs to
//! be satisfied in order to traverse that edge. For example:
//!
//! ```python
//! if cond:
//!   f()
//! else:
//!   g()
//! ```
//! Produces:
//! ```text
//!          +--------+
//!          |  Start |
//!          +--------+
//!              |
//!              v
//!           +-------+
//!   +------>|if stmt|
//!   |       +-------+
//!   |          |
//!   | cond     | not cond
//!   |          |
//!   v          v
//! +----+     +----+
//! | f()|     | g()|
//! +----+     +----+
//!   |          |
//!   +-----+----+
//!         |
//!       +--------+
//!       |  End   |
//!       +--------+
//!```
//!
//! ## Loops
//!
//! A loop consists of a _loop guard_, a _body_, and an optional
//! _else_ clause. The _loop guard_ is a condition we check to
//! determine whether to re-enter the loop body, the _body_ is again
//! a list of statements, and the _else_ clause is a possible
//! exit from the loop body or guard (depending). For simplicity,
//! let's ignore the else clause for a moment and specialize to the
//! case of `while` loops.
//!
//! When we reach a `while` loop, we begin by creating a new
//! basic block that is comprised entirely of the `loop` guard,
//! and which unconditionally follows the preceding basic block.
//! The loop guard consists of the test clause for the `while` statement.
//! This loop guard will have an outgoing node to the _loop exit_
//! and another to the loop body; the edge followed corresponds to the
//! veracity of the test clause.
//!
//! The loop body will almost always have two outgoing edges: one
//! which points back to the loop guard, and another that goes to the
//! loop exit. An exception would be the case of a jump statement, like
//! `continue` or `raise`.
//!
//! For example,
//!
//! ```python
//! while cond:
//!   continue
//! ```
//! would create
//!
//!```text
//!          ┌─────────┐  
//!          │         │  
//!          │  Start  │  
//!          │         │  
//!          └────┬────┘  
//!               │       
//!          ┌────▼────┐  
//!          │         │  
//!     ┌───►│  cond?  │  
//!     │    │         │  
//!     │    └─────┬──┬┘  
//!     │          │  │   
//!     │          │  │   
//! ┌───┴───────┐  │  │   
//! │           │  │  │   
//! │ continue  │◄─┘  │   
//! │           │     │   
//! └───────────┘     │   
//!                   │   
//!                   │   
//!           ┌───────▼──┐
//!           │          │
//!           │Loop Exit │
//!           │          │
//!           └──────────┘
//! ```
//!
//! ## Jumps
//!
//! Upon reaching a jump statement, we terminate the basic block
//! and add an outgoing edge. The target of the outgoing edge is
//! determined by the _loop context_ and the _try context_. That is,
//! we need to know whether and where we are within a try-statement,
//! and what (if any) is the innermost loop surrounding the current
//! statement.
//!
//! ### Break and Continue
//!
//! When we encounter a `break` or `continue`, we first check the
//! _try context_ to see if we are in a try context that has a
//! `finally` clause. In this case, we resolve the `finally` clause
//! first. Assuming no further jumps have occurred, we then check the
//! current loop context and direct flow to the loop exit (resp. guard)
//! in the case of a `break` (resp. `continue`).
//!
//! ### Raise and Return
//!
//! When we encounter a `raise` or `return`, we first check the
//! _try context_ stack to see if we are in a try context, and
//! whether it has an `except` or `finally` clause. In the case of
//! a `return`, we first resolve the `finally` block before returning.
//! In the case of a `raise`, we visit `except` blocks and then the
//! `finally` block. Without type inference, we are limited in our
//! ability to determine which `except` block is visited. As a first
//! pass, we pretend that we visit each of them.
//!
//! # Implementation Details
//!
//! Our implementation is non-recursive, guaranteed to terminate,
//! and takes a single, forward pass through a list of statements.
//!
//! To achieve this, our builder maintains two pieces of information:
//! - A stack of loop contexts
//! - A stack of `try` contexts
//!
//! # Misc other things
//! - When we encounter raise/return we should make remaining
//! things in local block an orphan node of some kind and mark
//! unreachable? Or add to worklist for later or something?
//! I guess there could be some context about being unreachable
//! or something...? Or maybe no special casing is required? Unclear.
pub mod builder;
pub mod implementations;
pub mod visualize;
