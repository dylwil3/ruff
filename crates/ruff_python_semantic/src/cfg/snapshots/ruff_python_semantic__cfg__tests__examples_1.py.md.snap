---
source: crates/ruff_python_semantic/src/cfg/mod.rs
description: "This is a Mermaid graph. You can use https://mermaid.live to visualize it as a diagram."
---
## Function 0
### Source
```python
def simple_unreachable():
    return 1
    return 2
    print("after")
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["return 1"]
	node1((("EXIT")))
	node2["return 2"]
	node3["print(#quot;after#quot;)"]
	node0==>node1
	node2==>node1
	node3==>node1
```

## Function 1
### Source
```python
def nested_if():
    if cond0:
        print("cond0")
        if cond1:
            print("cond1")
        elif cond2:
            print("cond2")
        else:
            print("cond0 else")
    else:
        print("else")
    print("after")
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["SWITCH"]
	node1((("EXIT")))
	node2["print(#quot;after#quot;)"]
	node3["print(#quot;cond0#quot;)
SWITCH"]
	node4["print(#quot;else#quot;)"]
	node5["print(#quot;cond1#quot;)"]
	node6["print(#quot;cond2#quot;)"]
	node7["print(#quot;cond0 else#quot;)"]
	node0-->|"cond0"|node3
	node0-->|"Else"|node4
	node2==>node1
	node3-->|"cond1"|node5
	node3-->|"cond2"|node6
	node3-->|"Else"|node7
	node4-->node2
	node5-->node2
	node6-->node2
	node7-->node2
```

## Function 2
### Source
```python
def nested_loops():
    for i in itr:
        while cond:
            print("inner")
        print("outer")
    print("done")
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["START"]
	node1((("EXIT")))
	node2["print(#quot;done#quot;)"]
	node3["LOOP GUARD"]
	node4["LOOP GUARD"]
	node5["print(#quot;outer#quot;)"]
	node6["print(#quot;inner#quot;)"]
	node0-->node3
	node2==>node1
	node3-->|"for i in itr"|node4
	node3-->|"Else"|node2
	node4-->|"cond"|node6
	node4-->|"Else"|node5
	node5-->node3
	node6-->node4
```

## Function 3
### Source
```python
def nested_try():
    try:
        try:
            foo()
        except:
            bar()
    except:
        buzz()
    finally:
        babble()
    print("done")
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["START"]
	node1((("EXIT")))
	node2["ENTER TRY"]
	node3["print(#quot;done#quot;)"]
	node4["EXCEPTION DISPATCH"]
	node5["babble()"]
	node6["RECOVERY"]
	node7["foo()"]
	node8["EXCEPTION DISPATCH"]
	node9["bar()"]
	node10["buzz()"]
	node0-->node2
	node2-->node7
	node3==>node1
	node4-->|"except any exception"|node10
	node4-->|"Else"|node5
	node5-->node6
	node6-->node3
	node7-->node8
	node8-->|"except any exception"|node9
	node8-->|"Else"|node4
	node9-->node4
	node10-->node5
```

## Function 4
### Source
```python
def nested_break():
    while cond:
        for i in itr:
            break
        print("outer")
        continue
    print("done")
```

### Control Flow Graph
```mermaid
flowchart TD
	node0["START"]
	node1((("EXIT")))
	node2["print(#quot;done#quot;)"]
	node3["LOOP GUARD"]
	node4["LOOP GUARD"]
	node5["print(#quot;outer#quot;)
continue"]
	node6["break"]
	node0-->node3
	node2==>node1
	node3-->|"cond"|node4
	node3-->|"Else"|node2
	node4-->|"for i in itr"|node6
	node4-->|"Else"|node5
	node5-->node3
	node6-->node5
```
